src/meta/greedy_load_balancer.cpp

```diff
@@ -903,102 +903,95 @@ void greedy_load_balancer::greedy_balancer(const bool balance_checker)
-                    return;
-                }
-            }
-        }
+    if (!execute_balance(apps,
+                         balance_checker,
+                         _balancer_in_turn,
+                         _only_move_primary,
+                         std::bind(&greedy_load_balancer::primary_balance,
+                                   this,
+                                   std::placeholders::_1,
+                                   std::placeholders::_2))) {
+        return;
     }
 
-    // TODO: do primary_balancer_globally when we find a good approach to
-    // make decision according to disk load.
-    // primary_balancer_globally();
-
-    if (!balance_checker) {
-        if (!t_migration_result->empty()) {
-            ddebug("stop to do secondary balance coz we already has actions to do");
-            return;
-        }
-    }
-    if (_only_primary_balancer) {
-        ddebug("stop to do secondary balancer coz it is not allowed");
+    if (!need_balance_secondaries(balance_checker)) {
         return;
     }
 
     // we seperate the primary/secondary balancer for 2 reasons:
     // 1. globally primary balancer may make secondary unbalanced
     // 2. in one-by-one mode, a secondary balance decision for an app may be prior than
     // another app's primary balancer if not seperated.
+    execute_balance(apps,
+                    balance_checker,
+                    _balancer_in_turn,
+                    _only_move_primary,
+                    std::bind(&greedy_load_balancer::copy_secondary,
+                              this,
+                              std::placeholders::_1,
+                              std::placeholders::_2));
+}
+
+bool greedy_load_balancer::need_balance_secondaries(bool balance_checker)
+{
+    if (!balance_checker && !t_migration_result->empty()) {
+        ddebug("stop to do secondary balance coz we already has actions to do");
+        return false;
+    }
+    if (_only_primary_balancer) {
+        ddebug("stop to do secondary balancer coz it is not allowed");
+        return false;
+    }
+    return true;
+}
+
+bool greedy_load_balancer::execute_balance(
+    const app_mapper &apps,
+    bool balance_checker,
+    bool balance_in_turn,
+    bool only_move_primary,
+    const std::function<bool(const std::shared_ptr<app_state> &, bool)> &callback)
+{
     for (const auto &kv : apps) {
         const std::shared_ptr<app_state> &app = kv.second;
         if (is_ignored_app(kv.first)) {
-            ddebug_f("skip to do secondary balance for the ignored app[{}]", app->get_logname());
+            ddebug_f("skip to do balance for the ignored app[{}]", app->get_logname());
             continue;
         }
-
         if (app->status != app_status::AS_AVAILABLE || app->is_bulk_loading || app->splitting())
             continue;
 
-        bool enough_information = copy_secondary_per_app(app);
+        bool enough_information = callback(app, only_move_primary);
```

Comment:
name `callback` is too general, name it `balance`?

Reply:
OK, I will change it in next pull request